File: src/co/nums/intellij/aem/htl/editor/braces/HtlBraceMatcher.java
package co.nums.intellij.aem.htl.editor.braces;

import co.nums.intellij.aem.htl.psi.HtlTokenTypes;
import com.intellij.lang.BracePair;
import com.intellij.lang.PairedBraceMatcher;
import com.intellij.psi.PsiFile;
import com.intellij.psi.tree.IElementType;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class HtlBraceMatcher implements PairedBraceMatcher {

	private static final BracePair[] PAIRS = {
			new BracePair(HtlTokenTypes.LEFT_BRACKET, HtlTokenTypes.RIGHT_BRACKET, false),
			new BracePair(HtlTokenTypes.LEFT_PARENTH, HtlTokenTypes.RIGHT_PARENTH, false)
	};

	@Override
	public BracePair[] getPairs() {
		return PAIRS;
	}

	@Override
	public boolean isPairedBracesAllowedBeforeType(@NotNull IElementType lbraceType,
			@Nullable IElementType contextType) {
		return true;
	}

	@Override
	public int getCodeConstructStart(PsiFile file, int openingBraceOffset) {
		return openingBraceOffset;
	}

}

File: src/main/kotlin/co/nums/intellij/aem/htl/editor/braces/HtlBraceMatcher.kt
package co.nums.intellij.aem.htl.editor.braces

import co.nums.intellij.aem.htl.psi.HtlTypes
import com.intellij.lang.*
import com.intellij.psi.PsiFile
import com.intellij.psi.tree.IElementType

private const val STRUCTURAL = true

class HtlBraceMatcher : PairedBraceMatcher {

    private val pairs = arrayOf(
            BracePair(HtlTypes.EXPR_START, HtlTypes.EXPR_END, !STRUCTURAL),
            BracePair(HtlTypes.LEFT_BRACKET, HtlTypes.RIGHT_BRACKET, !STRUCTURAL),
            BracePair(HtlTypes.LEFT_PARENTH, HtlTypes.RIGHT_PARENTH, !STRUCTURAL)
    )

    override fun getPairs() = pairs

    override fun isPairedBracesAllowedBeforeType(leftBraceType: IElementType, contextType: IElementType?) = true

    override fun getCodeConstructStart(file: PsiFile, openingBraceOffset: Int) = openingBraceOffset

}
